raiden/transfer/channel.py

```diff
@@ -811,6 +811,33 @@ def get_amount_locked(end_state: NettingChannelEndState) -> TokenAmount:
     return total_pending + total_unclaimed + total_unclaimed_onchain
 
 
+def get_batch_unlock_values(
+    channel_state: NettingChannelState,
+) -> Tuple[TokenAmount, TokenAmount]:
+    """Collect lock values for expired locks and onchain unlocked locks.
+    Returns:
+        tuple(TokenAmount, TokenAmount): (
+                                          value for unlocking our_state,
+                                          value unlocking partner_state
+                                         )
+    """
+    our_unlock_value = sum(
```

Comment:
This formula here is not correct.

unlocked, locked, and onchain unlocked are **disjoint** sets. So, it doesn't make sense to subtract the unlocked amount from the locked amount.

This is the meaning of these attributes:

- locked locks: Locks which have been introduced with a locked transfer, however the secret is not know yet.
- unlocked locks: Locks which the secret is known, but the partner has not unlocked it off chain yet.
- onchain unlock: Locks which the secret is known, the partner has not unlocked off chain, and the secret has been registered onchain before the lock has expired.

So, for the `unlock` call, the value that you can be received by a node is `onchain unlock`, the `locked` and offchain `unlock` goes back to the sender.

Another way to get these values, with a simpler formula, is to use the `locked_amount` from the balance proofs, which is the sum of all these locked amounts, and from it remove the `onchain unlock`.

Reply:
This is my suggestion for formulas:
```python
receive_our_side = our_bp.locked_amount - our_state.onchain_unlocked
receive_partner_side = partner_state.onchain_unlocked
```

Reply:
hmm. I see: so there are two issues:
1) we must consider "offchain unlocked" locks together with the "still locked" locks at `batch_unlock` time, because any locked amount that wasn't released by a new balance proof from the partner will count towards the locked amounts on contract side. Therefore any locked amount that is not on-chain unlocked will count towards the returned-to-sender amount.
2) my implementation was a "greedy" implementation: _only_ call `batch_unlock` for a channel end if we receive more than the partner from it. That is an unfortunate/incorrect heuristic for two reasons: 
a) even though the the partner would receive _more_ than us, we would still not get our remaining tokens from the contract, _if_ the partner does not unlock. This greedy behavior would still work for us given two healthy canonical clients (but fail on error conditions), but:
b) in case of a "stalemate" (expired locks amount equals on-chain unlocked amount), neither side would call `batch_unlock`. The token amounts would stay stuck in the contract.

Thanks @hackaugusto !

Reply:
> my implementation was a "greedy" implementation: only call batch_unlock for a channel end if we receive more than the partner from it. 

Why not just go with the simple approach of if we get **anything** from an unlock, do it?
This will probably lead to races of unlocks on-chain between us and the partner but that we should be able to handle gracefully.

It's not ideal as a solution but it's a stepping stone towards a more full-fledged solution (taking into account gas costs, token exchange rates e.t.c.) and is definitely better than what we currently have.